File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/linked_list/P876_MiddleOfTheLinkedList.py
# author: tobebetter9527
# since: 2023/10/31 9:10
from typing import Optional

from com.dsa.leetcode.labuladong.linked_list import ListNode


class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return None
        fast, slow = head, head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
        return slow

File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/linked_list/two_pointers/P876_MiddleOfTheLinkedList.java
package com.dsa.leetcode.labuladong.linked_list.two_pointers;

import com.dsa.leetcode.labuladong.linked_list.ListNode;

public class P876_MiddleOfTheLinkedList {
    public ListNode middleNode(ListNode head) {
        if (head == null) {
            return null;
        }
        ListNode fast = head;
        ListNode slow = head;
        while (fast != null && fast.next != null) {
            fast = fast.next.next;
            slow = slow.next;
        }
        return slow;
    }
}

